DM-47889: Prevent DB connection pool exhaustion in Butler server

dhirving opened the pull request:
Adjusted the database connection pool parameters and added a limit to the number of long-running queries that the server will execute simultaneously.

If a client tries to run a long-running query while the server is overloaded, the request will be rejected with an HTTP 503 and the client will retry later.  This also sets up the retry logic for HTTP 429, which will be used in the future for other kinds of rate limiting.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Increase Postgres connection pool size

Set Postgres connection pool size to match the FastAPI thread pool size.  Testing showed that performance degraded horrendously once we got into the "overflow" connections that were allowed by the previous configuration.

dhirving committed:
Limit the number of in-flight queries on the server

The server now tracks the number of streaming queries in progress, and rejects query requests that exceed a limit of 25 with an HTTP 503.  The client now retries on receiving a 503 or 429 with a Retry-After header.

This will allow the server to avoid exhausting its thread pool and database connection pool with long-running queries.

timj commented on python/lsst/daf/butler/remote_butler/server/handlers/_query_streaming.py:
```diff
@@ -43,11 +44,26 @@
+# How long we ask callers to wait before trying their query again.
+# The hope is that they will bounce to a less busy replica, so we don't want
+# them to wait too long.
+_QUERY_RETRY_SECONDS = 5
```
Are we planning to do randomized exponential backoff at some point in the future? Can we tell if this is the 100th time a client has attempted a retry?

dhirving replied:
We don't really want to go exponential (at least by default), since most consumers of the Butler can't tolerate a request hanging out indefinitely.  This is just supposed to buy us enough time for auto-scaling to get some more replicas booted up.  When this case is triggered the scarce resource is threads and database connections, so I'm not really worried about the minimal amount of CPU and network needed to tell the caller to go away once every 5 seconds.

I added a cap on the retry time on the client side.  Currently we don't have a way to track what a client is up to on the server side, but we will eventually have a redis instance (or Russ will) for more elaborate rate limiting.

dhirving committed:
Limit the maximum retry time for query streaming

If after a couple of minutes the server is still not able to handle our query, we should just bail so that a Butler server failure doesn't cascade indefinitely into other services trying to use the Butler.
